pkg/util/log/dedup.go

```diff
@@ -0,0 +1,272 @@
+	e.mu.Lock()
+	if !e.isActive || (e.isActive && time.Since(e.seenAt) < expireAfter) {
+		e.mu.Unlock()
+		return
+	}
+	keyValues := e.keyValues
+	if e.count > 1 {
+		keyValues = append(keyValues, dedupLogKey, e.count)
+	}
+	e.isActive = false
+	e.mu.Unlock()
+
+	// log entry after expiring
+	_ = next.Log(keyValues...)
+}
+
+func (e *dedupEntry) deactivate(next log.Logger) {
+	e.mu.Lock()
+	if !e.isActive {
+		e.mu.Unlock()
+		return
+	}
+	keyValues := e.keyValues
+	if e.count > 1 {
+		keyValues = append(keyValues, dedupLogKey, e.count)
+	}
+	e.isActive = false
+	e.mu.Unlock()
+
+	_ = next.Log(keyValues...)
+}
+
+// Deduper implement log.Logger, dedupes log lines.
+type Deduper struct {
+	callerPrefix       string
+	targetKeys         []string
+	maxDedupCount      int
+	expireEntriesAfter time.Duration
+	next               log.Logger
+	stopCh             chan chan struct{}
+	entries            []dedupEntry
+}
+
+// NewDedupLogger creates and initializes a new Deduper instance.
+func NewDedupLogger(callerPrefix string, targetKeys []string, next log.Logger) *Deduper {
+	return newDedupLogger(
+		callerPrefix,
+		targetKeys,
+		next,
+		maxEntries,
+		maxDedupCount,
+		expireEntriesAfter,
+		garbageCollectEvery,
+	)
+}
+
+func newDedupLogger(
+	callerPrefix string,
+	targetKeys []string,
+	next log.Logger,
+	maxEntries int,
+	maxDedupCount int,
+	expireEntriesAfter time.Duration,
+	garbageCollectEvery time.Duration,
+) *Deduper {
+	d := &Deduper{
+		callerPrefix:       callerPrefix,
+		targetKeys:         targetKeys,
+		maxDedupCount:      maxDedupCount,
+		expireEntriesAfter: expireEntriesAfter,
+		next:               next,
+		stopCh:             make(chan chan struct{}, 1),
+		entries:            make([]dedupEntry, maxEntries),
+	}
+	go d.loop(garbageCollectEvery)
+	return d
+}
+
+// Log implements log.Logger.
+func (d *Deduper) Log(keyvals ...interface{}) error {
+	if len(keyvals) == 0 {
+		return nil
+	}
+	if len(keyvals)%2 == 1 {
+		keyvals = append(keyvals, nil)
+	}
+	// check whether log entry matches caller prefix
+	if !d.hasCallerPrefix(keyvals) {
+		return d.next.Log(keyvals...)
+	}
+	// compute dedup key if target keys are present
+	dedupKey := computeDedupKey(keyvals, d.targetKeys)
+	if len(dedupKey) == 0 {
+		return d.next.Log(keyvals...)
+	}
+
+	// look for an active entry
+	for i := 0; i < len(d.entries); i++ {
+		if matched := d.entries[i].updateOrForward(keyvals, dedupKey, d.maxDedupCount, d.next); matched {
+			return nil
+		}
+	}
+	// if no active entry was found, look for an empty slot and register the new entry
+	for i := 0; i < len(d.entries); i++ {
+		if matched := d.entries[i].activate(keyvals, dedupKey); matched {
+			return nil
+		}
+	}
+	// no free slot available... forward entry
+	return d.next.Log(keyvals...)
+}
+
+// Stop stops deduper loop.
+// Invoking this method guarantees all pending entries to be forwarded after returning.
+func (d *Deduper) Stop() {
+	ch := make(chan struct{})
+	d.stopCh <- ch
+	<-ch
```

Comment:
that's an interesting way to wait for the `loop` routine to be stopped. 

I think a more common pattern would be to use a wait group and then in the `loop()` you'd do something like `defer <wait group>.Done()` at the beginning, that's would be a bit more generic. 

oth, i don't see why this solution wouldn't work too...

Reply:
only problem I see with the wait group approach is that would require having to define an extra field, right? but I agree, they are mostly similar.